tools/privacyidea-migrate-linotp.py

```diff
@@ -173,182 +220,135 @@
-
-        ins = resolver_table.insert().values(
-                name=name,
-                rtype=config.get("type")
-        )
-        res_id = conn_pi.execute(ins)
-        resolver_id = res_id.inserted_primary_key[0]
-        for resolver_entry in config.get("config"):
-            # save the list of resolver configuration
-            ins = resolver_config_table.insert().values(
-                resolver_id=resolver_id,
-                Key=resolver_entry.get("Key") or "",
-                Value=resolver_entry.get("Value") or "",
-                Type=resolver_entry.get("Type") or "",
-                Description=resolver_entry.get("Description") or ""
-            )
-            conn_pi.execute(ins)
-
-
-# Put resolvers into realms
-s = select([linotp_config_table])
-result = conn_linotp.execute(s)
-realms = {}
-for r in result:
-    config_key = r["Key"]
-    entry = config_key.split(".")
-    if entry[1] == "useridresolver" and entry[2] == "group":
-        resolvers = r["Value"].split()
-        realms[entry[3]] = [x.split(".")[3] for x in resolvers]
-
-print("Migrating realm definitions {0}".format(realms))
-
-# Save the resolvers in the realms
-"""For this we need to get the IDs of the realmname and the resolvernames"""
-for realmname, resolvers in realms.iteritems():
-    print("Migrating realm {0!s}".format(realmname))
-    s = select([realm_table]).where(realm_table.c.name == realmname)
+    s = select([resolver_table])
     result = conn_pi.execute(s)
     for r in result:
-        realm_id = r["id"]
+        resolver_id_map[r["name"]] = r["id"]
+
+    print("Realm-Map: {}".format(realm_id_map))
+    print("Resolver-Map: {}".format(resolver_id_map))
+
+# Process Tokens
+
+if MIGRATE.get("tokens"):
+    s = select([linotp_token_table])
+    result = conn_linotp.execute(s)
 
-    for resolvername in resolvers:
-        s = select([resolver_table]).where(resolver_table.c.name ==
-                                           resolvername)
+    i = 0
+    for r in result:
+        i = i + 1
+        print("processing token #{1!s}: {0!s}".format(r["LinOtpTokenSerialnumber"], i))
+        # Adapt type
+        type = r["LinOtpTokenType"]
+        if type.lower() == "hmac":
+            type = "HOTP"
+        # Adapt resolver
+        linotp_resolver = r["LinOtpIdResClass"].split(".")[-1]
+
+        # Adapt resolver type
+        resolver_type = ""
+        if r["LinOtpIdResClass"]:
+            resolver_type = r['LinOtpIdResClass'].split(".")[1]
+            resolver_type = resolver_map.get(resolver_type)
+
+        # Adapt tokeninfo
+        ti = {}
+        if r["LinOtpTokenInfo"]:
+            ti = json.loads(r["LinOtpTokenInfo"])
+
+        if MIGRATE.get("assignments"):
+            user_pin = r['LinOtpTokenPinUser']
+            user_pin_iv = r['LinOtpTokenPinUserIV']
+            # Map the LinOTP-Resolver to the PI-Resolver
+            resolver = ASSIGNMENTS.get("resolver").get(linotp_resolver)
```

Comment:
Would it be useful to check for ``resolver is None`` here? Because right now, I think the admin would not notice if they specified an incomplete mapping.

Reply:
I also thought we should add something like:

   CONFIGURES = False

and bail out, if it was not set to True, so that the admin is aware, that he needs to change data in the script.

Reply:
Sounds good to me!
But maybe it would still be useful to add a warning or even bail out in case the LinOTP resolver cannot be found in the mapping?

Reply:
I am currently not sure. I probably would prefer to only dump a warning.
It could be that there were a lot of resolver definitions, that you actually do not want to migrate.

So this would be a way to migrate the assignments to *resolver/realm1*, but not to *resolver/realm2*. Bailing out would be a bit... ...drastic.